keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,965 @@
+// that logger was initialized by the program and not by code that should
+// better receive a logger via its parameters. TODO can be used as a temporary
+// solution for such code.
+func Background() Logger {
+    return fallbackLogger
+}
+```
+
+#### logging helper API
+
+The following helper code will get copied from klog:
+- `ObjectRef`
+- `KMetaData`
+- `KObj`
+- `KRef`
+- `KObjs`
+
+To ensure that a single import is enough in source files, that package will also contain:
+
+```go
+type Logger = logr.Logger
+var NewContext = logr.NewContext
+// plus more as needed...
+```
+
+<<[UNRESOLVED @pohly ]>>
+
+TODO: this might be an opportunity to enhance performance of KObj without
+causing an API break in kobj - see https://github.com/kubernetes/klog/pull/261.
+
+<<[/UNRESOLVED]>>
+
+#### Logging in tests
+
+Ginkgo tests do not need to be changed. Because each process only runs one test
+at a time, the global default can be set and local Logger instances can be
+passed around just as in a normal binary.
+
+Unit tests with `go test` require a bit more work. Each test case must
+initialize a `logr/testing` Logger for its own instance of `testing.T`.
+
+<<[UNRESOLVED @pohly ]>>
+
+What is a good default log level for unit tests? Should JSON output be enabled
+by default in all unit tests that use the common helper code or be enabled by
+individual unit tests only if desired?
+
+<<[/UNRESOLVED]>>
+
+API:
+
+```go
+func NewTestContext(t *testing.T) (context.Context, Logger)
+```
+
+Example:
+
+```go
+import (
+    “testing”
+    “k8s.io/component-helpers/log”
+)
+
+func TestSomething(t *testing.T) {
+    // Either return value can be ignored with _ if not needed.
+    ctx, logger := log.NewTestContext(t)
+            logger.Info(“test starts”)
+            doSomething(ctx)
+}
+```
+
+Custom log helper code must use the `WithCallStackHelper` method to ensure that
+the helper gets skipped during stack unwinding:
+
+```go
+func logSomething(logger log.Logger, obj interface{}) {
+    helper, logger := logger.WithCallStackHelper()
+    helper()
+    logger.Info(“I’m just helping”, “obj”, obj)
+}
+```
+
+Skipping multiple stack levels at once via `WithCallDepth` is not working with
+loggers that output via `testing.T.Log`. `WithCallDepth` therefore should only
+be used by code in `test/e2e` where it can be assumed that the logger is not
+using `testing.T`.
+
+### [logcheck](https://github.com/kubernetes/klog/tree/main/hack/tools/logcheck)
+
+That tool is a linter for log calls. It’s used to check that log calls are
+well-formed. Because it is used in core Kubernetes and SIGs (for example,
+[metrics-server](https://github.com/kubernetes-sigs/metrics-server/blob/4b20c2d43e338d5df7fb530dc960e5d0753f7ab1/Makefile#L252-L257),
+it should better be hosted in a repo where “go install” is fast. It only
+depends on the standard Go runtime, therefore it can be moved to
+`k8s.io/klogr/logcheck`. The tool will be updated to check not only klog
```

Comment:
Are you planning any new checks? Otherwise, document in the KEP that no changes are planned for the tool.

Reply:
One change is already mentioned at the end of the paragraph:
> The tool will be updated to check not only klog calls, but also calls through the `logr.Logger` interface.

If there's consensus about not passing ctx and logger as parameters, that would also need to be mentioned here.

Reply:
FWIW, after integrating logcheck into golangci-lint, I am more comfortable with adding such a check because in those cases where both needs to be passed (for example, because of performance concerns), `nolint:logcheck` can be used to override the check.